internal/sdkprovider/kafkatopicrepository/repository_test.go

```diff
@@ -0,0 +1,285 @@
+				{topic: &aiven.KafkaTopic{TopicName: "b"}},
+				{topic: &aiven.KafkaTopic{TopicName: "c"}},
+				{topic: &aiven.KafkaTopic{TopicName: "a"}},
+				{topic: &aiven.KafkaTopic{TopicName: "b"}},
+				{topic: &aiven.KafkaTopic{TopicName: "c"}},
+				{topic: &aiven.KafkaTopic{TopicName: "a"}},
+				{topic: &aiven.KafkaTopic{TopicName: "b"}},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/a/a": {TopicName: "a"},
+				"a/a/b": {TopicName: "b"},
+				"a/a/c": {TopicName: "c"},
+				"a/b/a": {TopicName: "a"},
+				"a/b/b": {TopicName: "b"},
+				"a/b/c": {TopicName: "c"},
+				"b/a/a": {TopicName: "a"},
+				"b/a/b": {TopicName: "b"},
+			},
+			v1ListCalled: 3, // 3 different cervices
+			// 2 services has 3 topics each.
+			// Plus one service has 2 topics
+			// Gives us batches (brackets) with topics (in brackets):
+			// [2] + [1] + [2] + [1] + [2]
+			v2ListCalled:    5,
+			v2ListBatchSize: 2,
+		},
+		{
+			name: "v1List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("bla bla bla")},
+			},
+			v1ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    0,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: bla bla bla")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List 404 error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: topic list has changed")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       aiven.Error{Status: 404},
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+	}
+
+	for _, opt := range cases {
+		t.Run(opt.name, func(t *testing.T) {
+			client := &fakeTopicClient{
+				storage:   opt.storage,
+				v1ListErr: opt.v1ListErr,
+				v2ListErr: opt.v2ListErr,
+			}
+
+			ctx := context.Background()
+			rep := newRepository(client)
+			rep.workerCallInterval = time.Millisecond
+			rep.v2ListBatchSize = opt.v2ListBatchSize
+
+			// We must run all calls in parallel
+			// and then call the worker
+			var wg sync.WaitGroup
+			for i := range opt.requests {
+				wg.Add(1)
+				go func(i int) {
+					defer wg.Done()
+					topic, err := rep.Read(ctx, opt.requests[i].project, opt.requests[i].service, opt.requests[i].topic)
+					assert.Equal(t, opt.responses[i].topic, topic)
+					assert.Equal(t, opt.responses[i].err, err)
```

Comment:
the usage of `assert.Equal` for comparing errors may cause issues if the errors have different memory addresses but equivalent content; consider using `assert.ErrorIs` or comparing error strings directly

Reply:
`ErrorIs` compares error constant with returned error. I compare dynamic errors:

```
"topic read error: All attempts fail:\n#1: topic list has changed"
```

Is mixed errors from retry lib + repository error.

Reply:
in this case, I would recommend to use [`multierr`](https://github.com/uber-go/multierr) package to combine two errors together

Reply:
Why would I use third-party library if it just works as expected?

Reply:
because the usage of `assert.Equal` for comparing errors may cause issues if the errors have different memory addresses but equivalent content

Reply:
1. Can you prove this test doesn't work, please?
2. Could you share an official documentation guideline forcing to use `multierr` please?
3. Please take a look at the source code that merges all errors [into a string](https://github.com/avast/retry-go/blob/a9809a7cb7530870cd56b553d15aaeb87cabf1c3/retry.go#L253).

Reply:
instead of `multierr` you could use `errors.Join`, which is the suggested way according to the [official docs](https://go.dev/doc/go1.20#errors) as of Go v1.20, e.g. `err = errors.Join(err, nil, err2, err3)`

Reply:
Please answer the item 3).

Reply:
I wasn't saying the test doesn't works, what I was saying is that there is a possibility of flakiness due to different memory addresses in runtime

as for the point three, I don't understand exactly how are we dependent on the linked third party library in this specific scenario?

Reply:
The error is stringified. Therefore can't be detected by `errors.Is`.

Reply:
so, is it stringified in the `retry-go` library?

Reply:
upd: yes

Reply:
I have answered to this several times. And left a link to the code line.